{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "class ProxyPool(object):\n",
    "    def __init__(self):\n",
    "        self.url = ''\n",
    "        self.headers = {'User-Agent':'Mozilla/5.0'}\n",
    "        \n",
    "    def save_proxy(self):\n",
    "        html = requests.get(url=self.url,headers=self.headers).text\n",
    "        print(html)\n",
    "        ip_list = html.split('\\r\\n')\n",
    "        print(ip_list)\n",
    "        with open('/home/tarena/proxies.txt','a') as f:       \n",
    "            for ip in ip_list:\n",
    "                # 测试ip,可用的保存到文件\n",
    "                if self.test_ip(ip):\n",
    "                    # 把ip写入到文件\n",
    "                    f.write(ip+'\\n')\n",
    "    \n",
    "    def test_ip(self,ip):\n",
    "        proxies = {'http':'http://{}'.format(),'https':'https://{}'.format()}\n",
    "        try:\n",
    "            res = requests.get(url='http://www.baidu.com',proxies=proxies,timeout=2)\n",
    "            if res.status_code == 200:\n",
    "                \n",
    "                return True\n",
    "        except Exception as e:\n",
    "            print(ip,'不可用')\n",
    "            return False\n",
    "            \n",
    "        \n",
    "spider = ProxyPool()\n",
    "spider.save_proxy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109ac8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "60.166.148.173:17024 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a58>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "116.17.185.120:18306 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109828>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "123.169.34.226:22769 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109240>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "1.194.86.43:17512 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109e80>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "114.99.22.81:20581 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a20>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "60.169.52.216:15637 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109780>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "113.121.165.76:15801 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109860>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "112.91.78.131:18992 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109710>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "114.217.183.186:23525 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109b00>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "110.244.65.119:20183 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109da0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "115.229.202.199:20535 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109c18>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "113.222.82.229:15102 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c1099e8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "106.111.222.6:16761 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109be0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "1.199.40.191:23620 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109cf8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "60.175.213.132:21425 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109eb8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "113.121.44.124:15448 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109cc0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "218.87.239.74:21730 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109048>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "115.151.7.42:18458 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109320>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "140.250.189.78:20157 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109ac8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "125.123.65.130:22111 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109f28>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "144.123.68.175:22479 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c1099b0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "182.111.121.30:19090 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a58>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "115.203.22.229:17782 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109828>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "183.165.242.41:23733 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109240>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "59.59.214.97:19904 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109e80>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "27.31.101.198:22994 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a20>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "1.194.80.150:22492 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109780>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "175.155.214.145:18246 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109860>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "36.6.138.79:17593 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109710>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "171.80.186.187:21403 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109b00>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "49.71.135.173:16938 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109da0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "121.205.85.5:23658 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109c18>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "113.75.1.29:22617 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c1099e8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "115.221.127.217:17420 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109be0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "59.32.20.219:22386 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109cf8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "115.203.20.70:15512 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109eb8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "183.158.127.108:15078 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109cc0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "183.158.7.7:21556 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109048>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "125.87.102.194:21931 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109320>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "182.202.222.83:21022 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109ac8>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "175.4.146.132:21150 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109f28>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "117.67.246.178:18225 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c1099b0>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "59.32.46.91:23213 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a58>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "58.47.35.238:16022 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109828>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "222.89.87.70:16211 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109240>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "49.70.64.24:15731 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109e80>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "112.252.30.185:16134 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109a20>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "106.46.162.146:20010 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109780>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "163.179.206.181:20371 不可用\n",
      "HTTPConnectionPool(host='{}', port=80): Max retries exceeded with url: http://www.baidu.com/ (Caused by ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc99c109860>: Failed to establish a new connection: [Errno -2] Name or service not known',)))\n",
      "27.157.3.158:16245 不可用\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "# url = 'http://xicidaili.com/nn/'\n",
    "url = 'http://www.baidu.com'\n",
    "headers = {'User-Agent':'Mozilla/5.0'}\n",
    "\n",
    "proxies_list = []\n",
    "with open('/home/tarena/secret_ip.txt','r') as f:\n",
    "    for line in f:\n",
    "        proxies_list.append(f.readline().strip())\n",
    "# print(proxies_list)\n",
    "\n",
    "def test(proxy):\n",
    "    try:\n",
    "        res = requests.get(url=url,\n",
    "                        proxies=proxy,\n",
    "                        headers=headers,\n",
    "                        timeout=3\n",
    "               )\n",
    "        if res.status_code == 200:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "\n",
    "for proxy in proxies_list:\n",
    "    proxies = {\n",
    "        'http':'http://{}',\n",
    "        'https':'http://{}'\n",
    "    }\n",
    "    proxies['http'].format(proxy)\n",
    "    proxies['https'].format(proxy)\n",
    "    with open('/home/tarena/useable.txt','w') as f:\n",
    "        if test(proxies):\n",
    "            print(proxy,'可用')\n",
    "            f.write(proxy + '\\n')\n",
    "        else:\n",
    "            print(proxy,'不可用')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
